csource/csource.go

```diff
@@ -226,15 +239,55 @@ loop:
 					esc = append(esc, '\\', 'x', hex(v>>4), hex(v<<4>>4))
 				}
 				fmt.Fprintf(w, "\tNONFAILING(memcpy((void*)0x%x, \"%s\", %v));\n", addr, esc, size)
+			case prog.ExecArgCsum:
+				csum_kind := read()
+				csum := CsumInfo{Addr: addr, Kind: csum_kind, Chunks: make([]CsumChunk, 0)}
```

Comment:
You emit ExecInstrCopyin/ExecArgCsum instructions where they are not ready to be executed and in the wrong order, and then fix both things in executor and in csource. Moreover, executor and csource are bad places for complex logic.
I would be better to emit ExecInstrCopyin/ExecArgCsum instructions at the point where they are ready to be executed and in the correct order. And then make executor and csource simply follow the instruction stream. Then we will not need:

```
+type CsumChunk struct {
 +	Kind  uintptr
 +	Value uintptr
 +	Size  uintptr
 +}
 +
 +type CsumInfo struct {
 +	Addr   uintptr
 +	Kind   uintptr
 +	Chunks []CsumChunk
 +}
+struct csum_chunk {
 +	uint64_t kind;
 +	uint64_t value; // addr for arg_csum_chunk_data,
 +			// value for arg_csum_chunk_const
 +	uint64_t size;
 +};
 +
 +struct csum_info {
 +	char* addr;
 +	uint64_t kind;
 +	uint64_t chunks_num;
 +	struct csum_chunk chunks[MAX_CSUM_CHUNKS];
 +};
```

as we will just execute these instruction inline.